lib/datalad_repository.rb

```diff
@@ -0,0 +1,145 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# Implements a DataProvider that can get files from a Datalad Repository
+#
+# To explore a Datalad Repository, a separate cached version of the repo
+# with the git annex links is to be maintained separately so that we can
+# explore it to find the file metadate.
+#
+# This library will create that cache and use it separately to maintain
+# the file metadata and initiate any datalad or git annex calls that are
+# needed to facilitate the data provider capability
+
+require 'fileutils'
+
+class DataladRepository
+
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+  # Establish a connection and install the initial image of the datalad repo in a
+  # Temporary cache directory
+
+  def initialize(datalad_repository_url, datalad_relative_path,
+                 dp_id="", cr_id="")
+                 #local_repository_directory, local_cache_file_name)
+
+    @prefix = File.join(datalad_repository_url,datalad_relative_path)
+    @path_prefix = datalad_relative_path
+
+    # Here we will give a unique name to this repository's cache directory
+    # and assign it a Userfile to be able to maintain it, as this is related
+    # to a dataprovider, the dataprovider information can be used to build the
+    # directory name.
+    dp_id_here = dp_id.blank? ? "aaa" : dp_id
+    cr_id_here = cr_id.blank? ? "bbb" : cr_id
+    prefix_clean = @prefix.dup.tr(":/.","_")
+    @cache_dir_name = "Datalad.rr=#{cr_id_here}.dp=#{dp_id_here}.pre=#{prefix_clean}"
+    @cache_path = ""
+    self
+  end
+
+  def connected?
+    ### need to figure out how to tell if I can get stuff from datalad
+    cmd_string = "curl --head -f --connect-timeout 20 #{get_url.bash_escape} > /dev/null 2>>/dev/null"
+    system(cmd_string)
+  rescue
+      false
+  end
+
+  ####################################################################
+  # path manipulation methods
+  ####################################################################
+
+  def get_url(relative_path="")
+    relative_path == "" ? @prefix : File.join(@prefix,relative_path)
+  end
+
+  def get_full_cache_with_prefix(relative_path="")
+    relative_path == "" ? File.join(@cache_path,File.basename(@path_prefix))
+                        : File.join(@cache_path,File.basename(@path_prefix),relative_path)
+  end
+
+  ####################################################################
+  # datalad installation methods
+  ####################################################################
+
+  def install_repository
+   #raise "Datalad Repository at #{@prefix} cache_directory not set prior to install" if @cache_path.nil?
+    @cache_userfile = DataladSystemSubset.find_or_create_as_scratch(:name => @cache_dir_name) do |cache_dir|
+      system("
+             mkdir -p #{cache_dir.to_s.bash_escape}
+             cd #{cache_dir.to_s.bash_escape}
+             datalad install -r -s #{get_url.bash_escape}
```

Comment:
I think what we may want is this:

```
datalad install -s #{get_url.bash_escape} -r --recursion-limit 1
```

Reply:
On my test env I CTRL-C's the install because after 25 minutes it had already created 120,000 directories, even though this was supposed to be for the browse function of top level of datasets.datalad.org (which contains 23 entries)